### PR TITLE
OCD-3621: Add/Edit Optional Standards

### PR DESCRIPTION
**Pull request by andlar**

### CODE REVIEW

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SearchViewController.java [hunk 1]
```diff
@@ -994,6 +995,23 @@ private FuzzyChoices updateFuzzyChoices(FuzzyChoices fuzzyChoices)
         return result;
     }
 
+    @ApiOperation(value = "Get all possible optional standard options in the CHPL",
+            notes = "This is useful for knowing what values one might possibly search for.")
+    @RequestMapping(value = "/data/optional_standards", method = RequestMethod.GET,
```
Comment (kekey1) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/SearchViewController.java]: should we use the dash notation here for the URL? /data/optional-standards

**Review by kekey1 — CHANGES_REQUESTED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/certifiedproduct/service/CertificationResultService.java [hunk 2]
```diff
@@ -31,23 +36,28 @@
...
         this.certificationResultDetailsDAO = certificationResultDetailsDAO;
-        this.svapDao = svapDao;
+        this.svapDao = svapDAO;
+        this.optionalStandardDAO = optionalStandardDAO;
+        this.ff4j = ff4j;
     }
 
     public List<CertificationResult> getCertificationResults(CertifiedProductSearchDetails searchDetails) throws EntityRetrievalException {
         List<SvapCriteriaMap> svapCriteriaMap = svapDao.getAllSvapCriteriaMap();
+        List<OptionalStandardCriteriaMap> optionalStandardCriteriaMap = optionalStandardDAO.getAllOptionalStandardCriteriaMap();
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/certifiedproduct/service/CertificationResultService.java:57]: Something in the method name that's called here reads like it should be plural... getAllOptionalStandardCriteriaMaps()?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/edition2015/OptionalStandardReviewer.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    public OptionalStandardReviewer(ErrorMessageUtil msgUtil, FF4j ff4j) {
+        this.msgUtil = msgUtil;
+        this.ff4j = ff4j;
+    }
+
+    @Override
+    public void review(CertifiedProductSearchDetails listing) {
+        if (ff4j.check(FeatureList.OPTIONAL_STANDARDS)) {
+            listing.getCertificationResults().stream()
+            .filter(cert -> (cert.getTestStandards() != null && cert.getTestStandards().size() > 0))
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/edition2015/OptionalStandardReviewer.java]: I think I'm confused about this reviewer... I know optional standards are replacing test standards and there is currently some overlap, and I saw this work in your demo. But this is checking against testStandards only as far as I can tell. Where is the code putting optional standards in the testStandards array? I can't find it but am probably missing it. Should there also be validation against certResult.getOptionalStandards()? What about checking that the optional standard chosen is one of the ones allowed for the criteria (yes someone would have to use the API for this).

**Review by kekey1 — CHANGES_REQUESTED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/certifiedproduct/service/CertificationResultService.java (hunk 2 above)
Comment (andlar) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/certifiedproduct/service/CertificationResultService.java:57]: I can see that, but I was following the pattern of the SVAP map in the line above. I feel like changing one of them would imply I should change the other, and I don't know how far that might ripple

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/edition2015/OptionalStandardReviewer.java (hunk 3 above)
Comment (andlar) [reply chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/edition2015/OptionalStandardReviewer.java]: This one was poorly named. I renamed it as a separate reviewer that we might be able to delete later, and added a reviewer specifically for Optional Standards (including some tests for it)

**Review by kekey1 — DISMISSED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/OptionalStandardReviewer.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    errorMessageUtil.getMessage("listing.criteria.optionalStandard.invalidEdition",
+                            cr.getNumber(), getListingEdition(listing))));
+        } else {
+            List<CertificationResult> certificationResultsWithOptionalStandards = listing.getCertificationResults().stream()
+                    .filter(cr -> cr.isSuccess() && cr.getOptionalStandards() != null && cr.getOptionalStandards().size() > 0)
+                    .collect(Collectors.toList());
+
+            Map<Long, List<OptionalStandardCriteriaMap>> optionalStandardCriteriaMap = null;
+            try {
+            optionalStandardCriteriaMap = optionalStandardDAO.getAllOptionalStandardCriteriaMap().stream()
```
Comment (tmy1313) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/OptionalStandardReviewer.java]: Indent...

**Review by tmy1313 — DISMISSED** [review]

**Review by kekey1 — APPROVED** [review]

**Review by tmy1313 — APPROVED** [review]

**Pull request merged**